extensions/amp-skimlinks/0.1/amp-skimlinks.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CommonSignals} from '../../../src/common-signals';
+import {Services} from '../../../src/services';
+import {once} from '../../../src/utils/function';
+import {whenDocumentReady} from '../../../src/document-ready';
+
+import {Tracking} from './tracking';
+
+import {AffiliateLinkResolver} from './affiliate-link-resolver';
+import {SKIMLINKS_REWRITER_ID} from './constants';
+import {EVENTS as linkRewriterEvents} from './link-rewriter/constants';
+
+import {LinkRewriterManager} from './link-rewriter/link-rewriter-manager';
+import {Waypoint} from './waypoint';
+import {getAmpSkimlinksOptions} from './skim-options';
+import {getBoundFunction} from './utils';
+
+export class AmpSkimlinks extends AMP.BaseElement {
+  /**
+   * @param {!AmpElement} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @private {?../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = null;
+
+    /** @private {?../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = null;
+
+    /** @private {?../../../src/service/document-info-impl.DocumentInfoDef} */
+    this.docInfo_ = null;
+
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter-manager.LinkRewriterManager} */
+    this.linkRewriterService_ = null;
+
+    /** @private {?Object} */
+    this.skimOptions_ = null;
+
+    /** @private {?./tracking.Tracking} */
+    this.trackingService_ = null;
+
+    /** @private {?./affiliate-link-resolver.AffiliateLinkResolver} */
+    this.affiliateLinkResolver_ = null;
+
+    /** @private {?./waypoint.Waypoint} */
+    this.waypoint_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter.LinkRewriter} */
+    this.skimlinksLinkRewriter_ = null;
+
+    /** @private {?string} */
+    this.referrer_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.xhr_ = Services.xhrFor(this.win);
+    this.ampDoc_ = this.getAmpDoc();
+    this.docInfo_ = Services.documentInfoForDoc(this.ampDoc_);
+    this.viewer_ = Services.viewerForDoc(this.ampDoc_);
```

Comment:
you can move those above to contructor

Reply:
That was my initial attempt but it looks like `this.getAmpDoc()` can only be called after `buildCallback()`. https://github.com/ampproject/amphtml/blob/master/src/base-element.js#L246
 Since the other services depend on the `ampdoc` object, I can't move them to the constructor (unless I am missing something?).

Reply:
you can do `Services.documentInfoForDoc(this.element)`

@choumx is there a preference here? I think we want to have all class member fields declared in constructor right?